### PR TITLE
Resolve 'text ignores rotational part of transformation' (#698)

### PR DESCRIPTION
**Pull request by immaxchen**
## PR Summary

Considering the text rotation don't get transformed may be a desired behavior
(such as text for annotation to some data point). Here made it as an option to
control whether or not the text rotation get transformed.

- original behavior described in #698:
```python
text = plt.text(0, 0, 'hello', size=50, transform=R + ax.transData)
```
Screenshot from 2019-10-27 00-02-16-small

- behavior in this patch:
```python
text = plt.text(0, 0, 'hello', size=50, transform=R + ax.transData, 
                rotation_mode='anchor', rotation_transformed=True)
```
Screenshot from 2019-10-26 23-57-37-small

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/text.py [hunk 1]
```diff
@@ -1134,6 +1147,17 @@ def set_rotation(self, s):
         self._rotation = s
         self.stale = True
 
+    def set_rotation_transformed(self, t):
+        """
+        Set if the text rotation get transformed or not.
```
Comment (timhoffm) [inline lib/matplotlib/text.py]: Maybe better:
~~~
Whether rotations of the transform affect the text direction.
~~~

File: lib/matplotlib/text.py [hunk 2]
```diff
@@ -130,6 +130,7 @@ def __init__(self,
                  rotation_mode=None,
                  usetex=None,          # defaults to rcParams['text.usetex']
                  wrap=False,
+                 rotation_transformed=False,
```
Comment (timhoffm) [inline lib/matplotlib/text.py]: I'm struggling a bit with the name, maybe `transform_rotates_text`? Better suggestions welcome.

**Comment by anntzer** [comment]
Do we really want to expose a control knob as to whether rotation affects text?  It would seem(??) like we generally *do* want it to affect text.  Perhaps just have that for a transition period and warn if the transform has nonzero rotation but that rotation was getting ignored and won't be ignored after the transition?
(Not really sure of the best strategy here.)

**Comment by timhoffm** [comment]
> Do we really want to expose a control knob as to whether rotation affects text? It would seem(??) like we generally do want it to affect text.

Not sure how it works internally, but wouldn't that rotate tick labels on a polar plot?

**Comment by anntzer** [comment]
Ah, fair point.

**Comment by anntzer** [comment]
Does this need a whatsnew note?

**Comment by immaxchen** [comment]
Sure, no problem.

**Comment by QuLogic** [comment]
This needs a rebase to resolve conflicts.

**Comment by tacaswell** [comment]
@immaxchen I took the liberty of rebasing and force-pushing for you, I hope you do not mind.

**Comment by immaxchen** [comment]
sure & thank you for the help.

**Comment by anntzer** [comment]
To what extent would it be possible to make this also improve the situation with https://github.com/matplotlib/matplotlib/issues/17005?

**Comment by immaxchen** [comment]
#17005 can be easily fixed [this way](https://38295-1385122-gh.circle-artifacts.com/0/doc/build/html/gallery/text_labels_and_annotations/text_rotation_relative_to_line.html), not sure if you like it though.

File: lib/matplotlib/text.py [hunk 3]
```diff
@@ -228,7 +230,18 @@ def _get_multialignment(self):
 
     def get_rotation(self):
         """Return the text angle in degrees between 0 and 360."""
-        return get_rotation(self._rotation)  # string_or_number -> number
+        if self.get_transform_rotates_text():
+            angle = get_rotation(self._rotation)
+            x, y = self.get_unitless_position()
+            return self.get_transform().transform_angles([angle, ], [[x, y]])
```
Comment (tacaswell) [inline lib/matplotlib/text.py:236]: This seems like it has type instability?

Comment (immaxchen) [reply lib/matplotlib/text.py]: @tacaswell resolved, please have a look

File: examples/text_labels_and_annotations/text_rotation_relative_to_line.py [hunk 4]
```diff
@@ -10,8 +10,8 @@
 to something on the plot.  In this case, the correct angle won't be
 the angle of that object in the plot coordinate system, but the angle
 that that object APPEARS in the screen coordinate system.  This angle
-is found by transforming the angle from the plot to the screen
-coordinate system, as shown in the example below.
+can be provided by the helper parameter *transform_rotates_text*, as
+shown in the example below.
```
Comment (QuLogic) [inline examples/text_labels_and_annotations/text_rotation_relative_to_line.py:14]: (wrap as needed)
```suggestion
can be determined automatically by setting the parameter *transform_rotates_text*, as
shown in the example below.
```

File: lib/matplotlib/tests/test_text.py [hunk 5]
```diff
@@ -10,9 +10,9 @@
 from matplotlib.backend_bases import MouseEvent
 import matplotlib.patches as mpatches
 import matplotlib.pyplot as plt
+import matplotlib.transforms as mtransforms
 from matplotlib.testing.decorators import check_figures_equal, image_comparison
 
-
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_text.py:15]: Unnecessary change.

File: doc/users/next_whats_new/2019-12-22_transform-rotates-text-direction.rst [hunk 6]
```diff
@@ -0,0 +1,7 @@
+:orphan:
+
+Transform can rotates text direction
```
Comment (QuLogic) [inline doc/users/next_whats_new/2019-12-22_transform-rotates-text-direction.rst:3]:
```suggestion
Transform can rotate text direction
```

File: doc/users/next_whats_new/2019-12-22_transform-rotates-text-direction.rst [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+------------------------------------
+The new `.Text` parameter ``transform_rotates_text`` now allows to enable
```
Comment (QuLogic) [inline doc/users/next_whats_new/2019-12-22_transform-rotates-text-direction.rst:5]:
```suggestion
The new `.Text` parameter ``transform_rotates_text`` now sets whether
```
